File: lenses-alerts-plugin-api/src/main/java/io/lenses/alerting/plugin/javaapi/AlertingPlugin.java
/*
 * Copyright 2017-2019 Lenses.io Ltd
 */
package io.lenses.alerting.plugin.javaapi;

import io.lenses.alerting.plugin.javaapi.util.Try;

import java.util.List;
import java.util.Map;

public interface AlertingPlugin {

    /**
     * Provides the name for the alert integration service. For Slack integration this will read Slack.
     * @return The alert integration name
     */
    String name();

    /**
     * Provides information on what and how this alert integration service works.
     * @return
     */
    String description();

    /**
     * Provides a list of all the configuration keys
     * @return
     */
    List<ConfigEntry> configKeys();

    /**
     * Returns AlertingService for valid configuration (depends on the alerting service class).
     * @param config service configuration
     * @return alerting service or failure
     */
    Try<AlertingService> init(Map<String, String> config);
}

File: lenses-alerts-plugin-api/src/main/scala/io/lenses/alerting/plugin/scalaapi/AlertingPlugin.scala
/*
 * Copyright 2017-2019 Lenses.io Ltd
 */
package io.lenses.alerting.plugin.scalaapi

import io.lenses.alerting.plugin.javaapi.{AlertingPlugin => JAlertingPlugin}
import io.lenses.alerting.plugin.scalaapi.util.TryUtils._

import scala.jdk.CollectionConverters._
import scala.util.Try

/**
  * simple alerting plugin wrapper
  */
class AlertingPlugin(underlying: JAlertingPlugin) {

  def init(config: Map[String, String]): Try[AlertingService] = for {
    init <- Try {
      underlying.init(config.asJava)
    }
    service <- init.asScala
  } yield new AlertingService(service)

  def name: String = underlying.name()

  def description: String = underlying.description()

  def configKeys: Seq[ConfigEntry] = underlying.configKeys().asScala.map { c => ConfigEntry(c.getKey, c.getDescription) }.toList
}

object AlertingPlugin {
  def apply(underlying: JAlertingPlugin): AlertingPlugin = new AlertingPlugin(underlying)
}

case class ConfigEntry(key: String, desc: String)